### PR TITLE
Implement namespacing for doc comments IDs

### LINKED ISSUE #91759: rustdoc: "Namespace" user-written Markdown headers
Tracking issue for @jsha's idea described [here](https://rust-lang.zulipchat.com/#narrow/stream/266220-rustdoc/topic/DOM.20id.20tidy.20check/near/264504935).

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #91759.

I think it also completely removes the need for #86178 (yeay!).

Thanks to the namespacing, we can stop worrying about declaring the IDs used by rustdoc for its layout.

You can check it online [here](https://rustdoc.crud.net/imperio/doc-comment-id-namespace/std/index.html).

cc @jsha (since the idea comes from you :) )
r? @camelid

### CODE REVIEW

**Review by jsha — COMMENTED** [review]
Generally looks good! Note that `HeadingOffset` and `IdWrapping` are conceptually related - they both give markdown information about the immediate parent heading. So maybe combine them into one `ParentInfo` that contains both?

Instead of calling it `IdWrapping`, how about `IdPrefix`?

And instead of containing `Some(&' clean::Item)`, what about containing an `&'a str`, which is the prefix itself? That narrows down the information available through this and makes its purpose clearer.

**Comment by GuillaumeGomez** [comment]
> And instead of containing `Some(&' clean::Item)`, what about containing an `&'a str`, which is the prefix itself? That narrows down the information available through this and makes its purpose clearer.

That would force us to make this transformation everywhere, which doesn't seem like a good idea. I prefer when a same operation from multiple places is also done in one place (so the `format!("{}.{}", type, name)` here.

> Instead of calling it `IdWrapping`, how about `IdPrefix`?

Sounds good to me.

> Generally looks good! Note that `HeadingOffset` and `IdWrapping` are conceptually related - they both give markdown information about the immediate parent heading. So maybe combine them into one ParentInfo that contains both?

I'll take a look into that, maybe it could be shortened indeed!

**Comment by jsha** [comment]
> That would force us to make this transformation everywhere, which doesn't seem like a good idea. I prefer when a same operation from multiple places is also done in one place (so the format!("{}.{}", type, name) here.

What about this: `ParentInfo::new(&item, H3)`. Then inside the `new` constructor, `ParentInfo` calls `item.type_()` and stores only the `&'a str`, throwing away the rest of the `clean::Item`?

**Comment by camelid** [comment]
Hmm, I just realized that this will of course break all links that users have to their Markdown headings. Is that acceptable breakage? I feel like we should probably have more team discussion and an FCP.

**Comment by jsha** [comment]
One possibility - we might be able to write JS that detects a fragment that doesn't exist in the doc, and searches the actually-existing IDs for what it would have been.

So for instance if we get `#Example-3`, the JS would iterate over all ids. If it sees one ending in `.example`, it would bump a counter (`n`). When `n` reaches 3, the JS would target that ID and update the document fragment.

We probably wouldn't want to keep this forever, but we could keep it around for a couple of years so there's time to transition.

Also note that the existing links are not very stable. For instance adding a new method with its own `# Examples` would break links to all `#example-N` after it. Well, actually worse - it would silently renumber those. So a hyperlink that pointed at one place would unexpectedly point at a different place.

Also since old versions on docs.rs don't usually get regenerated, the amount of breakage in links to old versions will be minimal.

**Comment by camelid** [comment]
> We probably wouldn't want to keep this forever, but we could keep it around for a couple of years so there's time to transition.

Hmm, but then no one would transition 🙃 

I think it's okay to break the links to headings that are within methods since they're basically broken already. But I wonder if there's a way to mitigate the breakage for top-level (e.g., struct-level) headings since they're not already broken?

I'm going to tag this with relnotes since it should probably be in the compatibility notes, and I think we should do some more discussion on the issue and then have an FCP.

**Comment by jyn514** [comment]
@shepmaster you were saying this wasn't the solution you expected, right? Would it be a big deal to add a prefix for your markdown anchors / do you have a better solution in mind?

File: src/librustdoc/html/markdown.rs [hunk 1]
```diff
@@ -1,29 +1,6 @@
 //! Markdown formatting for rustdoc.
 //!
 //! This module implements markdown formatting through the pulldown-cmark library.
-//!
```
[the remark this answers is not in the document]
Comment (camelid) [reply src/librustdoc/html/markdown.rs:4]: `rustdoc::html::markdown` is used by the error-index-generator. Perhaps that's why these docs were here?

**Comment by jsha** [comment]
@GuillaumeGomez could you post a demo of the stdlib docs with this change so we can check various edge cases?

File: src/librustdoc/html/markdown.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/html/markdown.rs:4]: You are right, this is very surprising that it's done this way... (It's in `src/tools/error_index_generator`.) I wonder why it's not using the rustdoc flag directly instead.

**Comment by GuillaumeGomez** [comment]
Sure, here it is: https://rustdoc.crud.net/imperio/doc-comment-id-namespace/std/index.html

File: src/librustdoc/html/markdown.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (camelid) [reply src/librustdoc/html/markdown.rs:4]: That's a good point. I wonder if it needs to control some rustdoc internals?

**Comment by shepmaster** [comment]
> you were saying this wasn't the solution you expected, right?

I'm not sure that I fully knew what to expect 😉 

That said, I am surprised that it namespaces the user-supplied data. I guess I expected it to namespace the rustdoc-supplied data. That is, I was have expected this link:

image

to be `#rd.modules`. If I typed the markdown `# modules`, that would have gotten the link `#modules`.

> Would it be a big deal to add a prefix for your markdown anchors

That rustdoc adds or that I add?

**Comment by camelid** [comment]
> That said, I am surprised that it namespaces the user-supplied data. I guess I expected it to namespace the rustdoc-supplied data. That is, I expected this link:

See https://rust-lang.zulipchat.com/#narrow/stream/266220-rustdoc/topic/DOM.20id.20tidy.20check/near/264582439.

File: src/librustdoc/html/markdown.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/html/markdown.rs:4]: I'll try to change it to see what the result looks like.

[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/html/markdown.rs:4]: They provide their own style files. So better keep it as is for now.

**Comment by GuillaumeGomez** [comment]
Updated to keep the current error index generator working.

**Comment by GuillaumeGomez** [comment]
I also updated the GUI tests. However, considering how big the change is, I put it into its own commit.

**Comment by GuillaumeGomez** [comment]
Needs https://github.com/rust-lang/rust/pull/91834 to be merged first...

File: src/librustdoc/html/markdown.rs [hunk 2]
```diff
@@ -1430,68 +1445,70 @@ crate fn rust_code_blocks(md: &str, extra_info: &ExtraInfo<'_>) -> Vec<RustCodeB
     code_blocks
 }
 
+/// This wrapper exists only to allow the error_index_generator tool to use it.
+#[derive(Debug, Clone, Copy)]
+pub struct IdPrefix<'a>(IdPrefixInner<'a>);
+
+impl<'a> IdPrefix<'a> {
+    pub fn without_parent() -> Self {
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/html/markdown.rs:1453]: Indeed!

**Comment by GuillaumeGomez** [comment]
Updated!

**Comment by GuillaumeGomez** [comment]
So before I update all the broken anchors in the docs, I'll wait for confirmation the current code is good. ^^'

**Comment by GuillaumeGomez** [comment]
cc @jsha

**Comment by JohnCSimon** [comment]
@GuillaumeGomez 

Ping from triage: I'm closing this due to inactivity, Please reopen when you are ready to continue with this. 
Note: if you do please open the PR **BEFORE** you push to it, else you won't be able to reopen - this is a quirk of github. 
Thanks for your contribution.

@rustbot label: +S-inactive

**Pull request closed without merging**